src/spyglass/common/common_interval.py

```diff
@@ -209,6 +225,537 @@ def cleanup(self, dry_run=True):
+        """Finds the union (all times in one or both) for two interval lists
+
+        Parameters
+        ----------
+        interval_list1 : np.ndarray
+            The first interval list [start, stop]
+        interval_list2 : np.ndarray
+            The second interval list [start, stop]
+        min_length : float, optional
+            Minimum length of interval for inclusion in output, default 0.0
+        max_length : float, optional
+            Maximum length of interval for inclusion in output, default 1e10
+
+        Returns
+        -------
+        np.ndarray
+            Array of intervals [start, stop]
+        """
+        interval_list1 = self.times
+        interval_list2 = self._extract(other)
+
+        def _parallel_union(interval_list):
+            """Create a parallel list where 1 is start and -1 the end"""
+            interval_list = np.ravel(interval_list)
+            interval_list_start_end = np.ones(interval_list.shape)
+            interval_list_start_end[1::2] = -1
+            return interval_list, interval_list_start_end
+
+        il1, il1_start_end = _parallel_union(interval_list1)
+        il2, il2_start_end = _parallel_union(interval_list2)
+
+        # Concatenate the two lists so we can resort the intervals and apply the
+        # same sorting to the start-end arrays
+        combined_intervals = np.concatenate((il1, il2))
+        ss = np.concatenate((il1_start_end, il2_start_end))
+        sort_ind = np.argsort(combined_intervals)
+        combined_intervals = combined_intervals[sort_ind]
+
+        # a cumulative sum of 1 indicates the beginning of a joint interval; a
+        # cumulative sum of 0 indicates the end
+        union_starts = np.ravel(
+            np.array(np.where(np.cumsum(ss[sort_ind]) == 1))
+        )
+        union_stops = np.ravel(np.array(np.where(np.cumsum(ss[sort_ind]) == 0)))
+        union = [
+            [combined_intervals[start], combined_intervals[stop]]
+            for start, stop in zip(union_starts, union_stops)
+        ]
+
+        return Interval(np.asarray(union))
+
+    def censor(self, timestamps):
+        """Returns new interval list that starts/ends at first/last timestamp
+
+        Parameters
+        ----------
+        interval_list : numpy array of intervals [start, stop]
+            interval list from IntervalList valid times
+        timestamps : numpy array or list
+
+        Returns
+        -------
+        interval_list (numpy array of intervals [start, stop])
+        """
+        interval_list = self.times
+        # check that all timestamps are in the interval list
+        if len(self.contains(timestamps, as_indices=True)) != len(timestamps):
+            raise ValueError("Interval_list must contain all timestamps")
+
+        timestamps_interval = np.asarray([[timestamps[0], timestamps[-1]]])
+        return Interval(self._intersect(interval_list, timestamps_interval))
+
+    def set_difference_inds(self, other, reverse=False):
```

Comment:
What's the difference between this and complement?

Reply:
They were handling adjacent cases differently, but I managed to figure out a logic that can handle all my test cases